aws/ec2_test.go

```diff
@@ -0,0 +1,59 @@
+package aws
+
+import (
+	"testing"
+
+	awsgo "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/stretchr/testify/assert"
+)
+
+func createTestEC2Instance(t *testing.T, session *session.Session) ec2.Instance {
+	svc := ec2.New(session)
+
+	params := &ec2.RunInstancesInput{
+		ImageId:      awsgo.String("ami-e7527ed7"),
+		InstanceType: awsgo.String("t1.micro"),
+		MinCount:     awsgo.Int64(1),
+		MaxCount:     awsgo.Int64(1),
+	}
+
+	runResult, err := svc.RunInstances(params)
+	if err != nil {
+		assert.Failf(t, "Could not create test EC2 instance: %s", err.Error())
+	}
+
+	return *runResult.Instances[0]
+}
+
+func TestListInstances(t *testing.T) {
+	session, _ := session.NewSession(&awsgo.Config{
+		Region: awsgo.String("us-west-2")},
+	)
+
+	instance := createTestEC2Instance(t, session)
+	instanceEntries, err := getAllEc2Instances(session, "us-west-2")
+
+	if err != nil {
+		assert.Fail(t, "Unable to fetch list of EC2 Instances")
+	}
+
+	assert.NotEqual(t, 0, len(instanceEntries))
```

Comment:
We often run tests in parallel, and I'm afraid if some other tests happens to fire up an EC2 instance in the same account, this test will pass, even if the code isn't actually working. Consider tagging the instances you launch from the aws-nuke test and make sure the instance you find with `getAllEc2Instances` includes the tag you expect.

In fact, it's probably a good idea to tag ALL resources created by aws-nuke tests with "aws-nuke-test-<UNIQUE_ID>" so if aws-nuke's own tests leave resources around, we know what the cause is. We typically use this [UniqueId](https://github.com/gruntwork-io/terratest/blob/06be4e266055d4512a35d3069c43738133eb71b5/util/random.go#L24) function to generate unique IDs for all our tests.

Reply:
* This particular test line simply checks that the creation of an instance succeeded and the next line ensures that the instance created by the test is present in the returned list. If in the `TestNukeInstances` we ensure we only delete appropriately tagged instances then, I feel this line will be fine

Reply:
Consider the following:

1. We have two automated tests running in the same AWS account at the same time. The first test, `foo`, creates an EC2 instance. The second test, this one for `aws-nuke`, tries to create the EC2 Instance, but due to some bug in the test, fails.
1. Now you do your check, which is merely seeing that there are more than 0 instances created. The check will pass, as `foo` created an EC2 Instance, but it *should* fail, as the instance created by `aws-nuke` didn't actually succeed.

Reply:
Ah makes sense, thanks for the clarification. I've removed this line and I now specifically check for the presence of the instance that was created